### PR TITLE
Optional VCN

### PR DESCRIPTION
**Pull request by nlamirault**
See #462

### CODE REVIEW

**Comment by karthicgit** [comment]
@nlamirault  Thanks for the PR . May i know is it tested  and instead of getting the display name we can get the ocid directly.

**Comment by nlamirault** [comment]
If we take the `ocid` of the VCN as an input, that makes the solution less dynamic, doesn't it?
For example, in a CICD pipeline, we will create the VCN , then the OKE.
The VCN will have a different ocid with each creation, but always the same name.

**Comment by karthicgit** [comment]
If the vcn is also created from another terraform module you can pass the vcn ocid as an output to the oke module. Its better to use ocid as its unique rather than display name which is not.

**Comment by nlamirault** [comment]
Yes. I already do that, using the https://github.com/oracle-terraform-modules/terraform-oci-vcn module to manage the VCN.
But VCN and OKE Terraform code are as different components :

```
❯  tree terraform/oci/
 oci
├──  bastion
│   └──  terraform
│       ├──  backend-vars
│       │   └──  staging.tfvars
│       ├──  backend.tf
│       ├──  main.tf
│       ├──  outputs.tf
│       ├──  provider.tf
│       ├──  tfvars
│       │   └──  staging.tfvars
│       ├──  variables.tf
│       └──  vcn.tf
├──  oke
│   └──  terraform
│       ├──  backend-vars
│       │   └──  staging.tfvars
│       ├──  backend.tf
│       ├──  generated
│       │   └──  kubeconfig
│       ├──  main.tf
│       ├──  oke.tf
│       ├──  outputs.tf
│       ├──  provider.tf
│       ├──  tfvars
│       │   └──  staging.tfvars
│       └──  variables.tf
└──  vcn
    └──  terraform
        ├──  backend-vars
        │   └──  staging.tfvars
        ├──  backend.tf
        ├──  data.tf
        ├──  locals.tf
        ├──  main.tf
        ├──  outputs.tf
        ├──  provider.tf
        ├──  security_list.tf
        ├──  subnets.tf
        ├──  tfvars
        │   └──  staging.tfvars
        ├──  variables.tf
        └──  vcn.tf
```

If i  pass the ocid into the `staging.tfvars` of the oke subdirectory. 

Each time I create/destroy the VCN, I will have to make a commit to update the OCID of the VCN in the OKE component directory.
If I use the name, since it's always the same VCN name, no commit is necessary :)

**Comment by karthicgit** [comment]
@nlamirault May i know the terraform apply succeed for you for this change?

**Comment by nlamirault** [comment]
WIP actually. I've got a register timeout on workers creation ...

**Comment by hyder** [comment]
I agree with @karthicgit. I prefer we use the vcn's ocid as the display name is not unique. 

Otherwise, this is looking good. What if the VCN does not have IG, NAT and Service gateways? Or are you expecting the VCN creator to have created them already?

**Comment by karthicgit** [comment]
Regarding passing vcn ocid or displayname we can pass both using coalesce function and prefer the vcn ocid if not use the display name and get the details via datasource.

**Comment by karthicgit** [comment]
@nlamirault  May i request to share any timeline for this .

**Comment by JanisOrlovs** [comment]
There are needed upstream changes in VCN module, to support enable/disable. I have created them and posted as PR. 

Same should be done with other components. Plug and play solution, like this, is good for demos, but not serious for production deployments. I as an OCI TF user, don't want to go with the opinionated approach as I have got a lot of limitations included in our environments.  

For example, I don't need:
a) bastion
b) network
c) components

Everything is already is in our environment. In this case, I need fork your repo to make it usable for us. This undermines idea of public modules + reusability. One of the reasons, why I would not recommend anybody to use OCI (if possible)

**Comment by hyder** [comment]
Hi @JanisOrlovs.

Thanks for your feedback. We welcome and appreciate them.

You mentioned you've submitted a PR. However, I can't see any at the moment. Can you please share the link here?

You're right, this module provides an opinionated approach. It is useful for some customers, less so for others with different requirements and we don't believe in a 1-size-fits all either. The reusability question is something that is increasingly coming up so we are definitely thinking about how to best implement them.

I can understand your not wanting to use this module but I fail to see why this also leads you to recommending not using OCI. If there are other areas of frustration, please let us know so we can work to address them.

**Comment by karthicgit** [comment]
@nlamirault @hyder I have made few changes and did a basic test and its working .https://github.com/karthicgit/terraform-oci-oke/tree/reusevcn-4.1.6

**Comment by nlamirault** [comment]
@karthicgit Did you make any changes from my branch?

**Comment by karthicgit** [comment]
@nlamirault Yes. I made few small changes .

**Comment by nlamirault** [comment]
Thanks @karthicgit I update my PR.

**Comment by JanisOrlovs** [comment]
Hello @hyder 

By definition, as user, I expect to get reusable code pieces that I can use to build my opinionated approach. Your getting started modules are much less opinated, than mainline modules. IMHO it should be vice versa.

I am a consultant, for example, now I have reinvent the wheel  to deploy OKE cluster in existing env, with VCN's, with PAM's without any what comes out of the box her. With GKE it would take me 1-4 hours, with EKS same time. Here it's way more longer than I need to get it.

**Comment by karthicgit** [comment]
@nlamirault please update the document and terraform.tfvars.example. Hope you have tested it as well.

File: data.tf [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2017, 2021 Oracle Corporation and/or affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl
+
+data "oci_core_vcns" "these" {
+  count = var.create_vcn == true ? 0 : 1
```
Comment (karthicgit) [inline data.tf:5]: Do we need the count again as its already handled in locals ?

**Review by karthicgit — COMMENTED** [review]
please update the document and terraform.tfvars.example as well.

File: modules/oke/iam.tf [hunk 2]
```diff
@@ -41,5 +47,8 @@ resource "oci_identity_policy" "oke_volume_kms" {
   description    = "Policies for block volumes to access kms key"
   name           = var.label_prefix == "none" ? "oke-volume-kms" : "${var.label_prefix}-oke-volume-kms"
   statements     = local.oke_volume_kms_policy_statements
+  
+  freeform_tags = var.freeform_tags
```
Comment (karthicgit) [inline modules/oke/iam.tf:51]: May i know why this freeform tag is added for iam. It will assign the full map .

File: modules/oke/cluster.tf [hunk 3]
```diff
@@ -39,6 +39,8 @@ resource "oci_containerengine_cluster" "k8s_cluster" {
     }
   }
 
+  freeform_tags = var.freeform_tags["cluster"]
```
Comment (karthicgit) [inline modules/oke/cluster.tf:42]: Update terraform.tfvars.example with a sample value

File: modules/oke/nodepools.tf [hunk 4]
```diff
@@ -10,6 +10,8 @@ resource "oci_containerengine_node_pool" "nodepools" {
   kubernetes_version = var.cluster_kubernetes_version
   name               = var.label_prefix == "none" ? each.key : "${var.label_prefix}-${each.key}"
 
+  freeform_tags = var.freeform_tags["node_pool"]
```
Comment (karthicgit) [inline modules/oke/nodepools.tf:13]: Update terraform.tfvars.example with a sample  nodepool tag

File: variables.tf [hunk 5]
```diff
@@ -818,7 +864,7 @@ variable "service_account_cluster_role_binding" {
 # tagging
 variable "freeform_tags" {
   default = {
-    # vcn, bastion and operator tags are required
+    # vcn, bastion, and operator tags are required
```
Comment (karthicgit) [inline variables.tf:867]: comma is not required before and.

File: variables.tf [hunk 6]
```diff
@@ -90,6 +90,52 @@ variable "ssh_public_key_path" {
...
+variable "vcn_id" {
+  type        = string
+  description = "Existing VCN id where the resources will be created"
+  default     = ""
+
+}
+
+variable "ig_route_table_display_name" {
+  type        = string
+  description = "Name of the existing NAT Gateway route table"
```
Comment (karthicgit) [inline variables.tf:115]: please change description to Internet gateway

File: variables.tf [hunk 7]
```diff
@@ -90,6 +90,52 @@ variable "ssh_public_key_path" {
...
+variable "ig_route_table_id" {
+  type        = string
+  description = "Existing Internet Gateway Route table id"
+  default     = ""
+
+}
+
+variable "nat_route_table_display_name" {
+  type        = string
+  description = "Name of the existing Internet Gateway route table"
```
Comment (karthicgit) [inline variables.tf:128]: please change description to nat gateway.

**Comment by karthicgit** [comment]
@nlamirault  Let's keep the commit only for reuse vcn related changes . you can raise another PR for freeform_tags related changes. Is it ok ?

**Comment by hyder** [comment]
> Hello @hyder
> 
> By definition, as user, I expect to get reusable code pieces that I can use to build my opinionated approach. Your getting started modules are much less opinated, than mainline modules. IMHO it should be vice versa.
> 
> I am a consultant, for example, now I have reinvent the wheel to deploy OKE cluster in existing env, with VCN's, with PAM's without any what comes out of the box her. With GKE it would take me 1-4 hours, with EKS same time. Here it's way more longer than I need to get it.

@JanisOrlovs 

It would be more appropriate to debate this in discussions instead of a PR but your point is well taken. 

That being said, this is how we started this module. Some of the things you're looking for could not have been possible until recent improvement in Terraform itself e.g. conditional modules.

File: data.tf (hunk 1 above)
Comment (nlamirault) [reply data.tf:5]: I think so : 

```terraform
 vcn_id = var.create_vcn == true ? module.vcn[0].vcn_id : coalesce(var.vcn_id, data.oci_core_vcns.these[0].virtual_networks[0].id)
```

**Comment by nlamirault** [comment]
@hyder IMHO i think Terraform modules like [terraform-aws-modules](https://github.com/terraform-aws-modules) or [terraform-google-modules](https://github.com/terraform-google-modules) is a better structure.

**Comment by karthicgit** [comment]
@nlamirault   please apply the change for vcn_id ,nat_route_id to storage module in main.tf as well . Test will fail otherwise.  There is a terraform.tfvars.example in the root folder where the changes needs to be updated not under examples.Please update the doc as well .

**Comment by karthicgit** [comment]
@nlamirault please update the doc terraformoption.adoc with the new variables ,description ,default value etc and instructions.adoc create a new section for Reuse VCN and provide the instruction .. I have also commented about freeform_tags use in iam.tf . please remove it . Let me know if you need help.

**Comment by karthicgit** [comment]
@nlamirault please update the doc terraformoption.adoc with the new variables ,description ,default value etc and instructions.adoc create a new section for Reuse VCN and provide the instruction .. I have also commented about freeform_tags use in iam.tf . please remove it . Let me know if you need help.

File: datasource.tf [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2017, 2021 Oracle Corporation and/or affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl
+
+data "oci_core_vcns" "these" {
```
Comment (hyder) [inline datasource.tf:4]: Please rename the datasource to `vcns`.

**Review by hyder — CHANGES_REQUESTED** [review]
Minor changes for

- tags in the cluster (iam.tf)
- naming of the vcn data source

**Comment by hyder** [comment]
Can you also please create a new page under /docs for documenting how to implement the reuse vcn?

**Comment by hyder** [comment]
hi @nlamirault,

Can I please request you to sign the Oracle Contributor agreement: https://www.oracle.com/technical-resources/oracle-contributor-agreement.html ?

Once the minor changes are in and the contributor agreement signed, we'll be able to merge these changes. Thank you very much for your contribution.

**Comment by snafuz** [comment]
> @hyder IMHO i think Terraform modules like [terraform-aws-modules](https://github.com/terraform-aws-modules) or [terraform-google-modules](https://github.com/terraform-google-modules) is a better structure.

I can see the value in the way they have implemented their modules, particularly if you’re going to use it on top of an existing environment. 
This OCI module takes a different approach, still very valid, to provide a self-contained way to provision a full K8s environment from scratch. Of course to do that it ends up being quite opinionated.

With the features that are now available in terraform, I believe we can refactor it in order to provide both options. I would start making the submodules bastionsvc, network, oke and extension public and usable as stand-alone modules. Then top-level module will just be a wrapper that may be used as entry point to provision a full or partial environment based on the user requirements.

**Comment by hyder** [comment]
Ok thanks for your valuable feedback everyone. We'll plan for the refactoring and make the submodules public. 

@nlamirault There are a few actions pending.

**Comment by Djelibeybi** [comment]
@hyder FYI, the OCA for this PR has been submitted, processed and accepted.

**Comment by hyder** [comment]
> @hyder FYI, the OCA for this PR has been submitted, processed and accepted.

Thanks @Djelibeybi. We'll wait for @nlamirault to finish the pending changes before we merge.

**Comment by hyder** [comment]
hi @nlamirault, 

Will you also be able to add the documentation?

**Comment by nlamirault** [comment]
@hyder lack of time to write the documentation :(

**Comment by hyder** [comment]
Hi @nlamirault,

Your PR addresses a frequently requested feature. Until recently, we've not been able to do it. However, since you've completed it, we would very much like to merge your PR.

That being said, we need the documentation before we can merge. We are willing to wait until you can complete. 

I hope that's ok with you.

**Comment by 12345ieee** [comment]
This is an extremely important feature for me as well, could you @hyder list the doc files that need to be updated and I'll see if I can help @nlamirault out?

**Comment by karthicgit** [comment]
@12345ieee  i can help you with listing the doc changes. please update the doc terraformoption.adoc with the new variables ,description ,default value etc .Create a new section for Reuse VCN in docs and provide the instruction .please let me know if you need any more info.

**Comment by karthicgit** [comment]
@hyder  @12345ieee  I have updated the doc for reusevcn in my branch . please review. https://github.com/karthicgit/terraform-oci-oke/tree/optionalvcndoc

**Comment by karthicgit** [comment]
@nlamirault  could you please fetch the latest code and remove the freeform_tags in modules/oke/iam.tf .

**Review by hyder — APPROVED** [review]
Conditional approval. @karthicgit Please remove the freeform tags as we discussed.

**Pull request merged**